cpp/src/arrow/io/interfaces.h

```diff
@@ -343,5 +344,9 @@ ARROW_EXPORT
 Result<Iterator<std::shared_ptr<Buffer>>> MakeInputStreamIterator(
     std::shared_ptr<InputStream> stream, int64_t block_size);
 
+ARROW_EXPORT
+Result<AsyncGenerator<std::shared_ptr<Buffer>>> MakeRandomAccessFileGenerator(
```

Comment:
If this is beneficial for reading from S3, should it perhaps be called in S3 and GCS filesystem implementations of `OpenInputStream` instead of just the CSV datasets reader?
(does that sound doable/reasonable?)